code/net/ConnectX-5/mlx5.c

```diff
@@ -0,0 +1,509 @@
+    TN_DEBUG("Could not get a command_queues's physical address");
+    return false;
+  }
+
+  // Since CMDQ_PHY_ADDR_LOW is not 32 bits but 20, the lower 12 bits
+  // of the command_queues_phys_addr should be 0
+  if ((command_queues_phys_addr & BITS(0, 12)) != 0) {
+    TN_DEBUG("The physical address has not the lower 12 bits set to 0!");
+    return false;
+  }
+
+
+  reg_write(device.addr, 0x10, (uint32_t)(command_queues_phys_addr>>32));
+  // shift the command_queues_phys_addr to write only the top 20 bits of the bottom 32 bits
+  reg_write_field(device.addr, 0x14, REG_CMDQ_PHY_ADDR_LOW, ((uint32_t) (command_queues_phys_addr & 0x00000000FFFFFFFF)) >> 12);
+  reg_clear_field(device.addr, 0x14, REG_NIC_INTERFACE);
+  reg_clear_field(device.addr, 0x14, REG_LOG_CMDQ_SIZE);
+  reg_clear_field(device.addr, 0x14, REG_LOG_CMDQ_STRIDE);
+
+  /**
+   * Step 3.
+   * Read the initializing field from the initialization segment.
+   * Repeat until it is cleared (INIT_SEGMENT.initializing become 0).
+   * */
+  TN_VERBOSE("### Init - step 3");
+
+  IF_AFTER_TIMEOUT(INITIALIZING_TIMEOUT * 1000 * 1000,
+         !reg_is_field_cleared(device.addr, REG_INITIALIZING_OFFSET, REG_INITIALIZING)) {
+    TN_DEBUG("INIT_SEGMENT.initializing did not clear, cannot complete init setup");
+    return false;
+  }
+
+  /**
+   * Step 4.
+   * Execute ENABLE_HCA command.
+   * */
+  struct CommandQueueEntry commandQueueEntryEnableHCA;
+
+  memset(&commandQueueEntryEnableHCA, 0, sizeof(commandQueueEntryEnableHCA));  // init all fields with 0
+  commandQueueEntryEnableHCA.type = 0x07;  // Table 247
+  commandQueueEntryEnableHCA.input_length = 12;  // Length of ENABLE_HCA
+  commandQueueEntryEnableHCA.status = 1;  //Table 247.  SW should set to 1 when posting the command
+
+  struct Enable_HCA enableHca;
+  memset(&enableHca, 0, sizeof(enableHca));  // init all fields with 0
+  enableHca.opcode = 0x104; //23.1 Introduction - Table 1153
+
+  memcpy(&commandQueueEntryEnableHCA.command_input_inline_data, &enableHca, sizeof(enableHca));
+
+  // Check what was written in the command_input_inline_data
+  for (int i = 0; i < 4; ++i) {
+    TN_DEBUG("command_input_inline_data[%d]  = %x", i, commandQueueEntryEnableHCA.command_input_inline_data[i]);
+  }
+
+  // Writing the command entry to the command queue
+  memcpy(&command_queues_virt_addr, &commandQueueEntryEnableHCA, sizeof(commandQueueEntryEnableHCA));
+
+  // Check what was written to the command queue
+  uint32_t aux[16] = {};
+  memcpy(&aux, &command_queues_virt_addr, sizeof(commandQueueEntryEnableHCA));
+  for (int i = 0; i < 16; ++i) {
+    TN_DEBUG("Command queue byte[%d] = %x", i, aux[i]);
+  }
+
+  // Set the corresponding bit (0 for Enable_HCA as this is the first command in the queue)
+  // from command_doorbell_vector to 1
+  // OBS When driver is in No DRAM NIC mode, this field must not be written. TODO: should I check this ?
+  reg_write(device.addr, 0x18, 1);
```

Comment:
I have a segmentation fault here, but I do not know why.

I am trying to write 1 to the doorbell vector and this is a write only field according to the datasheet: 
https://imgur.com/a/ZVpkaP2

Reply:
In the `reg_write_raw` function the pointer is dereferenced and then assigned a value (`*reg_addr = tn_cpu_to_le32(value);`  ), maybe I am reading here a write only memory ? @SolalPirelli , what do you think ?

Reply:
`*x = y` only writes to x, without a read, so it should be fine. (If you used the `_field` variants of the `reg_` methods it'd read first)

Maybe it's this "No DRAM NIC mode" thing you added in a comment above?